src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -191,15 +311,113 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
     }
 
     /**
-     * Method to build field data
-     * @return String: comma separated list with leading space in square brackets
+     * Builds a field data string from a builder.
+     *
+     * @param builder                  aggregation or sort builder
+     * @param propertiesAsMap       properties
+     * @param index                 index
+     * @param showFieldName            whether to include field names
+     * @param showFieldType            whether to include field types
+     * @return Field data string
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    String buildFieldDataString(
+        NamedWriteable builder,
+        Map<String, Object> propertiesAsMap,
+        Index index,
+        Boolean showFieldName,
+        Boolean showFieldType
+    ) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            if (showFieldName) {
+                fieldDataList.add(fieldName);
+            }
+            if (showFieldType) {
+                String fieldType = getFieldType(fieldName, propertiesAsMap, index);
+                if (fieldType != null && !fieldType.isEmpty()) {
+                    fieldDataList.add(fieldType);
+                }
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    String getFieldType(String fieldName, Map<String, Object> propertiesAsMap, Index index) {
+        String fieldType = getFieldTypeFromCache(fieldName, index);
+        if (fieldType != null) {
+            if (fieldType.equals(NO_FIELD_TYPE_VALUE)) {
+                return null;
+            } else {
+                return fieldType;
+            }
```

Comment:
Why not simply return `NO_FIELD_TYPE_VALUE` ?

Reply:
Cannot do this, if fieldType is not null, i.e. we found the field in the cache. We can directly return fieldType though